cub/agent/agent_batch_memcpy.cuh

```diff
@@ -0,0 +1,967 @@
+ * this method with the same arguments
+ * @param dest Pointer to the memory location to copy to
+ * @param num_bytes Number of bytes to copy
+ * @param src Pointer to the memory location to copy from
+ */
+template <typename VectorT, typename ByteOffsetT>
+__device__ __forceinline__ void
+VectorizedCopy(int32_t thread_rank, int32_t group_size, void *dest, ByteOffsetT num_bytes, const void *src)
+{
+  char *out_ptr                      = reinterpret_cast<char *>(dest);
+  const char *in_ptr                 = reinterpret_cast<const char *>(src);
+  constexpr size_t out_datatype_size = sizeof(VectorT);
+  constexpr size_t in_datatype_size  = sizeof(uint32_t);
+
+  // Number of bytes between out_ptr and the first OutT-aligned address at or after out_ptr
+  auto const alignment_offset = reinterpret_cast<std::uintptr_t>(out_ptr) % out_datatype_size;
+
+  // The first OutT-aligned address before (or at) out_ptr
+  VectorT *out_chars_aligned = reinterpret_cast<VectorT *>(out_ptr - alignment_offset);
+
+  // Both `out_start_aligned` and `out_end_aligned` are indices into `out_ptr`.
+  // `out_start_aligned` is the first OutT-aligned memory location after `out_ptr + 4`.
+  // `out_end_aligned` is the last OutT-aligned memory location before `out_end - 4`, where out_end corresponds to one
+  // past the last byte to be copied. Bytes between `[out_start_aligned, out_end_aligned)` will be copied using
+  // uint4. `out_ptr + 4` and `out_end - 4` are used instead of `out_ptr` and `out_end` to avoid `LoadVector`
+  // reading beyond data boundaries.
+  int32_t out_start_aligned = (in_datatype_size + alignment_offset + out_datatype_size - 1) / out_datatype_size *
+                                out_datatype_size -
+                              alignment_offset;
+  int32_t out_end_aligned = (num_bytes - in_datatype_size + alignment_offset) / out_datatype_size * out_datatype_size -
+                            alignment_offset;
+
+  // Perform vectorized copies by loading 4B at the time to populate a `VectorT` and write that out to global memory
+  for (int64_t ichar = out_start_aligned + thread_rank * out_datatype_size; ichar < out_end_aligned;
+       ichar += group_size * out_datatype_size)
+  {
+    VectorT data_in;
+    LoadVector(in_ptr + ichar, data_in);
+    *(out_chars_aligned + (ichar + alignment_offset) / out_datatype_size) = data_in;
+  }
+
+  // Tail logic: copy bytes of the current string outside `[out_start_aligned, out_end_aligned)`.
+  if (out_end_aligned <= out_start_aligned)
+  {
+    // In this case, `[out_start_aligned, out_end_aligned)` is an empty set, and we copy the entire string.
+    for (int32_t ichar = thread_rank; ichar < num_bytes; ichar += group_size)
+    {
+      out_ptr[ichar] = in_ptr[ichar];
+    }
+  }
+  else
+  {
+    // Copy bytes in range `[out_start, out_start_aligned)`.
+    if (thread_rank < out_start_aligned)
+    {
+      out_ptr[thread_rank] = in_ptr[thread_rank];
```

Comment:
For some reasons I end up in this branch for completely aligned data. Is it expected?